lightning/src/ln/interactivetxs.rs

```diff
@@ -0,0 +1,1110 @@
+			(serial_id, input, tx)
+		}).collect();
+		let outputs_to_contribute = outputs_to_contribute.into_iter().map(|output| {
+			let serial_id = generate_local_serial_id(entropy_source, is_initiator);
+			(serial_id, output)
+		}).collect();
+		let mut constructor = Self {
+			state_machine,
+			channel_id,
+			inputs_to_contribute,
+			outputs_to_contribute,
+		};
+		let message_send = if is_initiator {
+			match constructor.do_local_state_transition() {
+				Ok(msg_send) => Some(msg_send),
+				Err(_) => {
+					debug_assert!(false, "We should always be able to start our state machine successfully");
+					None
+				}
+			}
+		} else {
+			None
+		};
+		(constructor, message_send)
+	}
+
+	fn do_local_state_transition(&mut self) -> Result<InteractiveTxMessageSend, AbortReason> {
+		// We first attempt to send inputs we want to add, then outputs. Once we are done sending
+		// them both, then we always send tx_complete.
+		if let Some((serial_id, input, prev_tx)) = self.inputs_to_contribute.pop() {
+			let msg = msgs::TxAddInput {
+				channel_id: self.channel_id,
+				serial_id,
+				prevtx: TransactionU16LenLimited(prev_tx),
+				prevtx_out: input.previous_output.vout,
+				sequence: input.sequence.to_consensus_u32(),
+			};
+			let _ = do_state_transition!(self, local_tx_add_input, &msg)?;
+			Ok(InteractiveTxMessageSend::TxAddInput(msg))
+		} else if let Some((serial_id, output)) = self.outputs_to_contribute.pop() {
+			let msg = msgs::TxAddOutput {
+				channel_id: self.channel_id,
+				serial_id,
+				sats: output.value,
+				script: output.script_pubkey,
+			};
+			let _ = do_state_transition!(self, local_tx_add_output, &msg)?;
+			Ok(InteractiveTxMessageSend::TxAddOutput(msg))
+		} else {
+			let msg = msgs::TxComplete { channel_id: self.channel_id };
+			let _ = do_state_transition!(self, local_tx_complete, &msg)?;
+			Ok(InteractiveTxMessageSend::TxComplete(msg))
+		}
+	}
+
+	pub fn handle_tx_add_input(&mut self, msg: &msgs::TxAddInput) -> Result<InteractiveTxMessageSend, AbortReason> {
+		let _ = do_state_transition!(self, remote_tx_add_input, msg)?;
+		self.do_local_state_transition()
+	}
+
+	pub fn handle_tx_remove_input(&mut self, msg: &msgs::TxRemoveInput) -> Result<InteractiveTxMessageSend, AbortReason> {
+		let _ = do_state_transition!(self, remote_tx_remove_input, msg)?;
+		self.do_local_state_transition()
+	}
+
+	pub fn handle_tx_add_output(&mut self, msg: &msgs::TxAddOutput) -> Result<InteractiveTxMessageSend, AbortReason> {
+		let _ = do_state_transition!(self, remote_tx_add_output, msg)?;
+		self.do_local_state_transition()
+	}
+
+	pub fn handle_tx_remove_output(&mut self, msg: &msgs::TxRemoveOutput) -> Result<InteractiveTxMessageSend, AbortReason> {
+		let _ = do_state_transition!(self, remote_tx_remove_output, msg)?;
+		self.do_local_state_transition()
+	}
+
+	pub fn handle_tx_complete(&mut self, msg: &msgs::TxComplete) -> Result<(Option<InteractiveTxMessageSend>, Option<Transaction>), AbortReason> {
```

Comment:
Can we also return the total value of the inputs contributed by the counterparty if we get a transaction out here? It's needed for the strict ordering of the sending of `tx_signatures` (which is always part of interactive construction but happens outside this state machine):

https://github.com/lightning/bolts/blob/6d51a8a53fb56910855958927636aeadaaf7c3ba/02-peer-protocol.md?plain=1#L403

```
The sending node:
  - if it has the lowest total satoshis contributed, as defined by total
    `tx_add_input` values, or both peers have contributed equal amounts
    but it has the lowest `node_id` (sorted lexicographically):
    - MUST transmit their `tx_signatures` first
```

We just need the total for the counterparty as we already have ours.

Reply:
Oh actually nevermind. I'll just add helper methods to InteractiveTxConstructor that can be called after the negotiation is complete that are relevant to all interactive tx construction sessions, like `should_send_tx_signatures_first()` etc.